### PR TITLE
benchmark

### PR DESCRIPTION
**Pull request by MXWXZ**

### CODE REVIEW

File: test/hook.lua [hunk 1]
```diff
@@ -1,6 +1,19 @@
...
+
 -- recv hook, m is message
-function recv(m)
-    print(m)
+function recv(path, m)
+    local code = 0
+    local file = ''
+    buf = buf..m
+    if (string.sub(buf, -4) == '\r\n\r\n')
+    then
```
Comment (tammela) [inline test/hook.lua]: `then` is usually on the same line as `if`

File: tls.h [hunk 2]
```diff
@@ -48,9 +48,9 @@
 #include <net/strparser.h>
 #include <crypto/aead.h>
 #include "uapi/tls.h"
-#include <lua.h>
-#include <lualib.h>
-#include <lauxlib.h>
+#include "lunatik/lua/lua.h"
```
Comment (tammela) [inline tls.h:51]: Why this change?

File: tls_main.c [hunk 3]
```diff
@@ -43,9 +43,9 @@
 
 #include "tls.h"
 
-#include <lua.h>
-#include <lualib.h>
-#include <lauxlib.h>
+#include "lunatik/lua/lua.h"
```
Comment (tammela) [inline tls_main.c:46]: Why this change?

File: test/hook.lua [hunk 4]
```diff
@@ -1,6 +1,19 @@
...
+        local method = string.match(buf, "%S+")
+        local name = string.match(buf, "%S+", string.len(method) + 2)
+        code = 200
+        file = path..name
```
Comment (tammela) [inline test/hook.lua]: You got the idea wrong here.
The correct idea is something like this:
```Lua
-- files that are available in the www root
-- use a set so setsockopt can manipulate this table by running `luaL_dostring()`
files = {}
files['foo.txt'] = true
files['boo.txt'] = true

function recv(msg)
   -- ...
   if method == "GET" and file[name] then
       return 200 -- the file is available
   end
end
```

Please adapt the PR accordingly

File: uapi/tls.h [hunk 5]
```diff
@@ -37,14 +37,17 @@
...
 #define TLS_LUA_RECVENTRY	4	/* Set lua parameters */
-#define TLS_LUA_ERRNO		5	/* Get lua errno */
+#define TLS_LUA_WWWROOT		5	/* Set lua wwwroot */
+#define TLS_LUA_ERRNO		10	/* Get lua errno */
+#define TLS_LUA_CODE		11	/* Get lua http code */
+#define TLS_LUA_FILE		12	/* Get lua file */
 
-#define TLS_LUA_OK			0	/* no error */
-#define TLS_LUA_RECVERR		1	/* lua recv func error */
+#define TLS_LUA_OK 0 /* no error */
```
Comment (tammela) [inline uapi/tls.h]: Style changes here, what happened?

**Review by tammela — CHANGES_REQUESTED** [review]

File: tls.h (hunk 2 above)
Comment (MXWXZ) [reply tls.h:51]: For the case that system lua header version is different from lunatik. Mine is the same but I'm not sure others with low or newer version can work

File: test/test_server.c [hunk 6]
```diff
@@ -125,6 +121,69 @@ char *read_file(char *path)
...
+	char file[4096];
+	int maxsize;
+	int intlen;
+	char *http_header;
+	int fd;
+	struct stat file_stat;
+	maxsize = sizeof(file);
+	intlen = sizeof(int);
+
+	getsockopt(client, SOL_TLS, TLS_LUA_CODE, &code, &intlen);
```
Comment (VictorNogueiraRio) [inline test/test_server.c]: Do you need to look at the code if you already have the file in the cache?

File: tls_main.c [hunk 7]
```diff
@@ -662,6 +694,11 @@ static int do_tls_setsockopt(struct sock *sk, int optname, char __user *optval,
 		rc = do_tls_lua_setrecv(sk, optval, optlen);
 		release_sock(sk);
 		break;
+	case TLS_LUA_WWWROOT:
```
Comment (VictorNogueiraRio) [inline tls_main.c]: Why not load the `WWWROOT` value inside a Lua script? I believe doing so would really simplify your code.

File: tls_main.c [hunk 8]
```diff
@@ -424,15 +425,31 @@ static int do_tls_getsockopt_tx(struct sock *sk, char __user *optval,
...
+	switch (optname) {
+	case TLS_LUA_ERRNO:
+		if (val_len != sizeof(int))
+			return -EINVAL;
+		copy_to_user(optval, &ctx->lua_err, sizeof(int));
+		break;
+	case TLS_LUA_CODE:
+		if (val_len != sizeof(int))
+			return -EINVAL;
+		copy_to_user(optval, &ctx->lua_wwwcode, sizeof(int));
```
Comment (VictorNogueiraRio) [inline tls_main.c:443]: Maybe `http_replycode` would be better. @tammela what do you think?

File: test/hook.lua [hunk 9]
```diff
@@ -1,6 +1,19 @@
+buf = ''
```
Comment (VictorNogueiraRio) [inline test/hook.lua:1]: you could do something like:
```
buf = ''
wwwroot = 'randomdir'
files = {'index.html', 'test.html', ... }
```

Or load the `wwwroot` value and the files in another script.

**Review by VictorNogueiraRio — CHANGES_REQUESTED** [review]

File: test/hook.lua [hunk 10]
```diff
@@ -1,6 +1,23 @@
...
+
 -- recv hook, m is message
-function recv(m)
-    print(m)
+function recv(path, m)
+    local code = 0
+    local file = ''
+    buf = buf..m
+    if (string.sub(buf, -4) == '\r\n\r\n') then
+        local method = string.match(buf, "%S+")
```
Comment (VictorNogueiraRio) [inline test/hook.lua]: Maybe this would be simpler:
```
local name = buf:match('GET /(%S+) HTTP/%d%.%d\r\n')
```
You can assume that, if you can't match GET, the method is incorrect.

ps: untested code!
pps: Instead of `/` you should use `path's` value and, also, take a look at this: https://tools.ietf.org/html/rfc3986#section-3.3 to create a more precise pattern

**Review by VictorNogueiraRio — CHANGES_REQUESTED** [review]

File: test/test_server.c (hunk 6 above)
Comment (MXWXZ) [reply test/test_server.c]: The request parser is handled in lua script so the c server didn't know the file name requested or other information.
